services/QuillLMS/spec/queries/admin_diagnostic_reports/diagnostic_performance_by_skill_query_spec.rb

```diff
@@ -0,0 +1,149 @@
+          Array.new(questions.length - correct_count) { |i| create(:concept_result, activity_session: activity_session, question_number: i + 1 + correct_count, extra_metadata: {question_uid: questions[i + correct_count].uid}, correct: false) }
+        ].flatten
+      end
+
+      let(:concept_results) do
+        [
+          pre_activity_sessions.map { |activity_session| concept_results_for_activity_session(activity_session, pre_questions, pre_correct_count) },
+          post_activity_sessions.map { |activity_session| concept_results_for_activity_session(activity_session, post_questions, post_correct_count) }
+        ]
+      end
+
+      let(:diagnostic_id) { pre_diagnostic.id }
+
+      let(:query_args) do
+        {
+          timeframe_start: timeframe_start,
+          timeframe_end: timeframe_end,
+          school_ids: school_ids,
+          grades: grades,
+          teacher_ids: teacher_ids,
+          classroom_ids: classroom_ids,
+          aggregation: aggregation_arg,
+          diagnostic_id: diagnostic_id
+        }
+      end
+
+      # Some of our tests include activity_sessions having NULL in its timestamps so we need a version that has timestamps with datetime data in them so that WITH in the CTE understands the data type expected
+      let(:reference_activity_session) { create(:activity_session, :finished) }
+
+      let(:cte_records) do
+        [
+          classrooms,
+          teachers,
+          students,
+          classrooms_teachers,
+          schools,
+          schools_users,
+          classroom_units,
+          pre_activity_sessions,
+          post_activity_sessions,
+          concept_results,
+          pre_diagnostic,
+          post_diagnostic,
+          reference_activity_session,
+          questions,
+          diagnostic_question_skills,
+          skill_group
+        ]
+      end
+
+      context 'invalid diagnostic id' do
+        let(:diagnostic_id) { 1 }
+
+        it { expect{ subject }.to raise_error(DiagnosticPerformanceBySkillQuery::InvalidDiagnosticIdError) }
+      end
+
+      # Ideally we'd break this up into multiple cases, and also account for more varied inputs, but for some reason this query takes 15-ish seconds to run even with this dummy data, so writing a bunch of separate cases would balloon our run time.  Writing one super-case for now.
+      # TODO: Come back and expand the test cases in this spec once they can be run in reasonable time
+      it 'maximal results test' do
+        result = results.first
+        all_pre_concept_results = pre_activity_sessions.map(&:concept_results).flatten
+        all_post_concept_results = post_activity_sessions.map(&:concept_results).flatten
+
+        expect(result[:aggregate_rows].length).to eq(classrooms.map(&:grade).uniq.length)
+        expect(result[:skill_name]).to eq(skill_group.name)
+        expect(result[:pre_score]).to eq(all_pre_concept_results.select { |cr| cr.correct }.length.to_f / all_pre_concept_results.length * 100)
```

Comment:
I think the following four `select { |cr| cr.correct }` can be rewritten `select(&:correct)`